core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/WorkflowWorker.scala

```diff
@@ -64,7 +66,46 @@ class WorkflowWorker(
     logManager.sendCommitted
   )
   val timerService = new WorkerTimerService(actorService)
-  val dpThread = new DPThread(actorId, dp, logManager, inputQueue)
+
+  val dpThread =
+    new DPThread(actorId, dp, logManager, inputQueue)
+
+  def setupReplay(): Unit = {
+    if (workerConf.replayTo.isDefined) {
+
+      context.parent ! ReplayStatusUpdate(actorId, status = true)
+
+      val (processSteps, messages) = ReplayLogGenerator.generate(logStorage)
+      val replayTo = workerConf.replayTo.get
+      val onReplayComplete = () => {
+        logger.info("replay completed!")
+        context.parent ! ReplayStatusUpdate(actorId, status = false)
+      }
+      val orderEnforcer = new ReplayOrderEnforcer(
+        logManager,
+        processSteps,
+        startStep = logManager.getStep,
+        replayTo,
+        onReplayComplete
+      )
+      dp.inputGateway.addEnforcer(orderEnforcer)
+      messages.foreach(message => dp.inputGateway.acceptMessage(message))
+
+      logger.info(
+        s"setting up replay, " +
+          s"current step = ${logManager.getStep} " +
+          s"target step = ${workerConf.replayTo.get} " +
+          s"# of log record to replay = ${messages.size}"
+      )
+    }
+  }
+
+  override def initState(): Unit = {
+    dp.InitTimerService(timerService)
+    setupReplay()
+    dp.initOperator(workerIndex, workerLayer, currentOutputIterator = Iterator.empty)
```

Comment:
is the order correct? maybe first init operator then replay?

Reply:
changed.